### ISSUE TITLE
Extend pubgrub to match cargo's dependency resolution

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust-project-goals#110]
| Metadata         | |
| --------         | --- |
| Point of contact | @eh2406 |
| Team(s)          | [cargo](https://github.com/rust-lang/cargo) |
| Goal document    | [2025h1/pubgrub-in-cargo](https://rust-lang.github.io/rust-project-goals/2025h1/pubgrub-in-cargo.html) |

## Summary

Implement a standalone library based on pubgrub that model's cargo dependency resolution and bring it to a quality of code so that it can be maintained by the cargo team. This lays the groundwork for improved cargo error messages, extensions for hotly requested features (e.g., better MSRV support, CVE-aware resolution, etc), and support for a richer ecosystem of cargo extensions.

## Tasks and status

* [ ] Discussion and moral support ([cargo](https://github.com/rust-lang/cargo) ![Team][])
* [ ] Implementation work on pubgrub library ()

[Team]: https://img.shields.io/badge/Team%20ask-red

**Comment by nikomatsakis** [comment]
This issue is intended for status updates only.

For general questions or comments, please contact the owner(s) directly.

**Comment by Eh2406** [comment]
Almost daily updates are being posted on https://rust-lang.zulipchat.com/#narrow/stream/260232-t-cargo.2FPubGrub/topic/Progress.20report questions or feedback are absolutely welcome on that zulip stream.

**Comment by Eh2406** [comment]
Since the writing of this project goal we now have achieved for all crate versions on crates.io the two resolvers accept the other one's solution. If cargo thinks there is a valid lock file for a crate version then we check that:
1. The new PubGrub based resolver also thinks there is a valid solution for that crate version.
2. If cargo resolver is loaded with only the packages PubGrub selected, cargo still finds a solution.
3. If the PubGrub resolver  is loaded with only the packages cargo resolver selected,  PubGrub still finds a solution.
Otherwise, when cargo thinks there is no valid solution, then the new PubGrub based resolver concurs. This effort involved fixing many bugs related to whether optional dependencies have been activated. This milestone was achieved much faster than expected!

In addition, significant progress has been made speeding up running these tests. Over 30% improvements to the average performance to the new PubGrub based resolver, and important changes to cargo resolver to better allow it to be run in parallel. Big thank you to the Cargo Team for reviewing and merging these changes to the existing resolver. Running resolution in parallel is important to working on this project goal, but not useful to cargo.

**Comment by Eh2406** [comment]
There were some footnotes to last updates headline achievement "for all crate versions on crates.io the two resolvers accept the other one's solution", which all have now been addressed.  There were some versions filtered out because the existing cargo resolver did not accept the record from crates.io. @tareknaser look into these cases and determined that filtering them out was the correct behavior, they also do not work when using cargo to import them as a dependency. Big thanks to him for the contribution! After resolution is complete cargo checks for cyclical dependencies. The new resolver was augmented to also have this check. Tests ensure that this behavior exactly matches between resolvers.

Performance has made another massive improvement. The new resolver is on average faster when processing all of crates.io. This is also true when including the handful of pathological crates. These pathological crates used to hit hardcoded timeouts and are now each under two minutes. All of crates.io (with a 160s timeout) took 4.30hr and now takes 2.98hr, ~0.30% improvement.

Most runs of cargo happen with lock file. Lock file performance needed a lot of improvement, and got it. Verifying the lock file of nonpathological cases on crates.io went from 61.44min to 44.90min, a ~27% improvement. (The lock files of pathological cases verify very quickly, I just happened not to record those numbers.) There are several approaches for closing the remaining gap with the existing resolver.

**Comment by Eh2406** [comment]
Work on this project is pivoted to trying to make improvements to the underlying pubgrub-rs library in light of the work on this goal. We found a 10% performance win for checking lock files. By betting that things are going to work smoothly until the first time we have to backtrack. When checking the lock file, or if we are getting lucky, complications will not occur so it is a waste of effort to avoid them.

A long-standing issue with pubgrub-rs is an inefficiency when handling situations where several versions of a package are unavailable. The inefficiency leaks out as a error message with an explanation of each version that's unavailable. Instead we would prefer an error message with one line explaining that all the versions are unavailable for the same reason. Several attempts been made to fix this issue, all unsuccessful. One possible reason may have to do with a poorly understood change in behavior introduced early in the project. I read through all history of the project discussions to better understand what was known about the change. I created branches from old versions of the code to exactly identify changes in behavior. Now that we know how to bring back the original behavior we have to determine if it is better. By investigating the unsuccessful fix on top of the old behavior I was able to figure out why the fix did not work.

Meanwhile the tool has had some important quality of life improvements. Clap is now used to control what parts of the experiment actually need to be run. Useful when a change can only affect some of the behavior. Also it now works on Windows, and thanks to mimalloc with reasonable multithreaded performance. Performance improvements were made to reading the index and preparing crates as a resolution problem, time I spend waiting that is not part of the reported benchmark numbers.

**Comment by Eh2406** [comment]
The biggest progress has been contributions by @x-hgg-x https://github.com/rust-lang/cargo/pull/14583 improving the resolver test suite in Cargo to check feature unification against a SAT solver. This was followed up by https://github.com/rust-lang/cargo/pull/14614 porting the test cases that tripped up PubGrub to Cargo's test suite. This lays the groundwork so when Cargo switches to PubGrub we do not regress on these important behaviors. It also prepares for Fuzzing of features in dependency resolution.

**Comment by Eh2406** [comment]
We have published one of PubGrubs  core data structures as a separate [version-ranges](https://crates.io/crates/version-ranges) crate. Allowing several projects to share this  core abstraction, and to  utilize new improvements without waiting for the rest the project. This is one of many things required to publish a new `0.3.0`  version of the PubGrub crate.

**Comment by Eh2406** [comment]
We have landed several important speedups. The slowest crate now takes 11sec to resolve, [just a few updates ago](https://github.com/rust-lang/rust-project-goals/issues/110#issuecomment-2305818050) it was taking over 120sec. Checking all crates is now just 71.42min, down from 178min. Verifying the lock file of all crates is now 32.77min, down from 44.90min (without several of the hardest cases). We also landed a performance improvement to the existing resolver, a small change to how one data structure was hashed.

I've been keeping notes in this thread about how long it took to run some version of my script against some version of the index. Over time my program has gotten faster, but the index is gotten bigger. I've been looking into trying to pull these two factors apart. I was able to construct a graph of how the same code handle different versions of the index going back several months.

<img src="https://rust-lang.zulipchat.com/user_uploads/4715/wqcLIHz_HhT1WELaAAFFLgrm/image.png">

Going the other way, I looked at to runs  on the same index from different versions of my benchmarking code.

 | Wall | Pub Lock | Cargo Lock | Cargo | Pub
---|---|---|---|---|---
Jul 12 2024 | 101.27 | 145.02 | 64.85 | 670.4 | 472.19
Dec 17 2024 | 14.47 | 39.95 | 57.25 | 464.43 | 83.16 
% Improved | 86% | 72% | 12% | 31% | 82% 

Looking at only the so-called "pathological"  cases we see a 90% improvement.
Graphing this data  gives the dramatic plot:

<img src="https://rust-lang.zulipchat.com/user_uploads/4715/_319rRlGu74b_MXRk0SuAp0i/image.png">

We could go back further,  but the "random commit" I chose was a correctness fix. In comparing the performance of something that works correctly with something that doesn't is somewhat apples to oranges. Although there was already a significant performance work being done before then.

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h1

**Comment by Eh2406** [comment]
The major update so far is the release of [PubGrub 0.3](https://github.com/pubgrub-rs/pubgrub/releases/tag/v0.3.0). This makes available the critical improvements made to allow the functionality and performance demanded by Cargo and UV. The other production users can now take advantage of the past few years of improvements. Big thanks to @konstin for making the release happen.

Other progress is been stymied by being sick with Covid for a week in January and the resulting brain fog, followed by several high-priority projects for $DAY_JOB. It is unclear when the demands of  work will allow me to  return focus to this project.

**Comment by Eh2406** [comment]
Progress continues to be stalled by high priority tasks for $DAY_JOB. It continues to be unclear when the demands of work will allow me to return focus to this project.

**Comment by Eh2406** [comment]
I will be giving a talk at Rust-Week about the history that brought us to this project/goal. Aside from preparing for that I have not had time for this effort.

**Comment by Eh2406** [comment]
The talk went smoothly and was well received. I had several useful and interesting conversations at Rust Week about  effort. That is all I have to report.

**Comment by Eh2406** [comment]
My time at Amazon is coming to an end. They supported the very successful effort with the 2024h2 goal, and encouraged me to propose the 2025h1 goal that is now wrapping up. Unfortunately other work efforts led to the very limited progress on the 2025h1 goal. I do not know what comes next, but it definitely involves taking time to relax and recover. Recovering involves rediscovering the joy in the work that I love. And, I have a deep passion for this problem. I hope to make some time to work on this. But, relaxing requires reducing the commitments I have made to others and the associated stress. So I will not promise progress, nor will I renew the goal for 2025h2.
